crates/aptos-crypto/src/bls12381/bls12381_sigs.rs

```diff
@@ -0,0 +1,147 @@
+
+//! This module provides APIs for aggregating and verifying BLS multi-signatures
+//! implemented on top of BLS12-381 elliptic curves (https://github.com/supranational/blst).
+//!
+//! The `Signature` struct is used to represent either a signature share from an individual
+//! signer or a multisignature aggregated from many such signers.
+//!
+//! The signature verification APIs in `Signature::verify` and `Signature::verify_arbitrary_msg` do NOT
+//! assume the signature to be a valid group element and will implicitly "group-check" it. This
+//! makes the caller's job easier and, more importantly, makes the library safer to use.
+
+use std::convert::TryFrom;
+use anyhow::{anyhow, Result};
+use blst::BLST_ERROR;
+use serde::Serialize;
+use aptos_crypto_derive::{DeserializeKey, SerializeKey};
+use crate::bls12381::bls12381_keys::{PublicKey, PrivateKey};
+use crate::hash::CryptoHash;
+use crate::{CryptoMaterialError, Length, signing_message, ValidCryptoMaterial, ValidCryptoMaterialStringExt};
+use crate::bls12381::DST_BLS_MULTISIG_IN_G2_WITH_POP;
+use crate::traits;
+
+#[derive(Debug, Clone, Eq, PartialEq, SerializeKey, DeserializeKey)]
+/// Either A BLS signature share from an individual signer or a BLS multisignature aggregate from
+/// multiple such signers
+pub struct Signature {
+    pub(crate) sig: blst::min_pk::Signature,
+}
+
+////////////////////////////////////////////////
+// Implementation of (multi)signature structs //
+////////////////////////////////////////////////
+
+impl Signature {
+    /// The length of a serialized Signature struct
+    // NOTE: We have to hardcode this here because there is no library-defined constant
+    pub const LENGTH: usize = 96;
+
+    /// Serialize a Signature.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.sig.to_bytes()
+    }
+
+    /// Group-checks the signature: i.e., verifies the signature is a valid group element.
+    /// WARNING: This is called implicitly when verifying the signature in this struct's
+    /// `Signature::verify_arbitrary_msg` trait implementation. Therefore, this function should not
+    /// be called separately for most use-cases. We leave it here just in case.
+    pub fn group_check(&self) -> Result<()> {
+        self.sig.validate(true).map_err(|e| anyhow!("{:?}", e))
+    }
+
+    /// Optimistically-aggregate multiple signatures. The individual signature shares are not
+    /// necessarily validated nor necessarily group-checked. For performance reasons, we do not
+    /// group-check the signature shares here, since the verification of the returned multisignature
+    /// includes such a group check. As a result, adversarial signature shares cannot lead to
+    /// forgeries.
+    // CRYPTONOTE(Alin): I am wondering if it's safe to aggregate without group checking the
+    // signature shares. Specifically, maybe there's an attack where one could craft a signature
+    // share outside of the group that makes the aggregated multisignature verify.
```

Comment:
Are CRYPTONOTE docstrings stripped off from public exposure? I am staring at this comment and thinking maybe we shouldn't let others read this.

Reply:
I would say no, though I could be wrong. Such notes are important for ourselves to remember what we did, but also for the community who might audit the code. This note happens to be redundant since the comments above it summarize it, so I will remove it.